backend: lower modules into procedures

zerbina opened the pull request:
## Summary

- turn module-level imperative code into a procedure (a module's 'init'
  procedure)
- generate the program entry-point in a mostly backend-agnostic way
- change the destruction order of globals, and make it the same across
  backend
- make initialization behaviour of (module-level) globals the same
  across all backend
- make module initialization order the same across all backends (except
  the IC backend)

Almost all decision making regarding initialization order of globals and
modules is moved out of the code generators, and into a central place,
simplifying the code generators and making it much easier to adjust this
behaviour in the future.

Notable changes in behaviour:
- the correct module initialization order is now the one previously used
  with the C backend
- "lifted" globals (those marked with the `.global` pragma) for a module
  are now always initialized before the module-level ones 
- globals are destroyed on a per-module basis; modules are destroyed in
  the reverse order they were initialized (i.e. their module-level code
  run) in
- the `header` pragma is now respected for globals that don't have an
  initial value specified
- `asm` and `emit` statement cannot be used for emitting top-level
  JavaScript code anymore

## Details

### Lowering a module

Turning a module's imperative code into a procedure is integrated into
the collect pass (`setupModule`); once the full semanticized AST for a
module is present, the parts of the declarative AST that the code
generators still need is wrapped in a standalone statement list, while
the imperative AST is pre-processed and wrapped in a procedure.

Pre-processing means:
- scanning the module-level AST for definitions of globals and
  threadvars, and registering them with the `Module` instance
- changing the owner of all symbols defined in the imperative AST to be
  the wrapped-in procedure

The `var` and `let` sections with the globals cannot be turned into
assignments during the pre-processing. This is because:
- `nkForStmt`s weren't lowered yet
- the `injectdestructors` still needs to know about the definition
  points of globals in nested scopes
- the semantics of an "initializing assignment" cannot be encoded with
  `PNode` AST

So instead, transforming the definitions into assignment is done during
the MIR phase (more on that later). Before wrapping up setup of a
`Module` instance, symbols for all other "module-bound" operators are
created. These are:
- module-data initialization procedure
- module de-init (a.k.a. destructor) procedure
- module pre-init procedure

The module destructor is already filled with destructor calls for
globals defined in the outermost scope.

### MIR phase

Transforming the definitions of globals is implemented as two MIR passes
(`rewriteGlobalDefs`). They're run as follows:
- rewrite definitions of globals defined in the *outermost* scope
- `injectdestructors` pass
- rewrite definitions of globals defined in *nested* scopes

With handling of destruction for globals defined in the outermost scope
now being the responsibility of the earlier module processing,
`injectdestructors` only has to defer destructor calls (the
`globalDestructors` mechanism) for lifted globals. The only reason the
`globalDestructors` still needs to exists is that is that `jsgen`
doesn't yet extract lifted globals prior to running the MIR passes. 

After all MIR passes have run, no definitions for globals are left in
the AST that is output for the init procedures -- they were all replaced
with assignments.

### Code generation

The new `backends` module contains code-generator-agnostic routines that
are used by all orchestrators.

The changes to the code generators / orchestrators can be summarized as:
- emitting / handling the definitions for globals is now the
  responsibility of the orchestrators
- deciding on the content of the 'main' procedure is the responsibility
  of the orchestrators (which delegate this to `generateMain`,
  `generateTeardown`, and `generateMainProcedure`)
- the code generators are no longer responsible for handling destruction
  of globals

It's only possible to generate the code for destroying lifted globals
once all alive ones were discovered, but this is only the case after the
main part of code generation is done, as only then has all alive code
been discovered. For this reason, the orchestrators are currently
responsible for calling `finishDeinit` at the correct point in time.

Additional information and highlighted changes.

#### C backend

`cgen` still needs to append code to a module's init procedure from
outside of the init procedure, meaning that the module-init procedures
still have to be special cased.

The code generation logic around globals simplifies a bit, and
`cgenWriteModules` now no longer alters the C code for a module. Because
'main' procedure generation is handled elsewhere, the call to
`initThreadVarsEmulation` is appended to the end of the `system`
module's data-init procedure.

#### JS backend

A special stack-trace entry is inserted for module-level code when using
the JS backend. In order to keep this behaviour, the init procedure is
marked with a symbol flag (the repurposed `sfMainModule`) that signals
to `genProcBody` that the special stack-trace entry is to be used.

#### VM backend

A lot of the code in `vmbackend` was dedicated to module-level code and
entry-point generation, so the module shrinks down quite a bit.

- re-use the "module closed" order information from `ModuleList`
- use the correct module context when generating code
- remove dedicated support for module-level code generation
- replace the manual entry-point code generation with enqueuing the
  procedure produced by `generateMainProcedure`

#### IC backend

In order to make the backend unification process easier, the IC backend
is turned into a frontend for `backend/cbackend.nim`. It mostly acts as
a rodfile -> `ModuleList` adapter now, but still performs the alive
analysis.

A negative side-effect of this is, that when using the IC backend, the
code generator is now also invoked for modules of which neither the
content nor the set of alive symbols changed, significantly increasing
compile times when nothing changed.

### Other changes

- adjust the JavaScript implementation of `std/sysrand`, now that `emit`
  cannot be used to emit top-level JavaScript code anymore
- mark the `globalsSlot` global in `threadlocalstorage` as `.noInit`,
  which is necessary now that all globals without an initial value are
  default-initialized by the module's init procedure

### Changes to tests

- move the test from `tclosure_issues.nim` into a `knownIssue` test; it
  now fails with all backends
- change the `js/tclosure.nim` test to not rely on `asm` producing top-
  level JavaScript code
- don't import modules unavailable for the current platform in
  `tstyle_imports` and `tstrict_funcs_imports` (necessary now that the
  `header` pragma on globals works properly)
- enable some global-related tests for the VM target

saem commented:
OMG, this so good. Making it a proc would make it so so good.

zerbina commented:
I rebased the PR onto the `devel` branch, which fixed the closure-iterator-related issues. In addition, I've also rewritten the commit history, making the commits less broad and easier to follow along.

During the second bootstrap iteration, the compiler crashes with an access violation error inside one of procedures related to `refc` -- a `devel` compiler is able to build the compiler without any problems. An update of the csources compiler is probably needed now.

zerbina committed:
collectors: first steps

Implement the basic logic for lowering modules into procedures and
objects (structs), plus the data types for storing the results.

zerbina committed:
collectors: generate the initial module destructors

The initial module destructors can already be generated when setting up
the module, as the modules' globals are already known at that point.

zerbina committed:
cbackend: move main procedure generation out of `cgen`

How the main procedure looks like and in what order the modules are
initialized is now decided in a central location: `backends` (the
JavaScript and VM backends still need to be adjusted).

zerbina committed:
cbackend: setup globals and threadvars

With their definitions removed during generation of a module's `init`
procedure, emitting the C definitions is now the responsibility of the
orchestrator.

zerbina committed:
jsbackend: move main procedure generation out of `jsgen`

For now, a hack is used in order to keep the special stack-trace entry
for what was previously top-level code.

zerbina committed:
vmbackend: use `backends`

zerbina committed:
doc: adjust the manual

Remove the part stating that pure globals are initialized after any
top-level variables in their originating module. This was only
sometimes true prior to the changes from this PR, and now it's never
the case.

zerbina committed:
injectdestructor: adjust to the global variable changes

Checking for threadvars and globals defined in a module's outermost scope
is no longer necessary, as their definition statements are removed by the
MIR pass running prior to the `injectdestructors` pass.

We only need to check for pure globals now, but once those are properly
extracted when using the JS backend, the last-remaining special-casing
for globals can be removed.

zerbina committed:
jsbackend: emit definitions for globals

zerbina committed:
collectors: remove global definitions during the MIR phase

The previous approach to handling globals had a few issues:
- the implementation of the pre-pass was complex
- for-vars weren't extracted, and implementing this is effectively not
  possible prior to `transf`
- there's no way to express than an assignment to an external location
  is the initializing one with `PNode` AST. That's a problem when
  turning the definitions for globals into assignments, as the "first
  assignment" information is lost, causing the `injectdestructors` pass
  to insert `=sink` calls instead of blit copies for the definition-
  rewritten-to-assignment statements

To get around all this, replacing the definitions is moved to the MIR
phase (much simpler to implement), with `setupModule` only scanning the
AST, but not modifying it.

The `def`s for globals in nested scopes still need to be present for the
`injectdestructors` pass, so rewriting the defs happens in two steps:
- before destructor injection: rewrite only defs in the outermost scope
- after destructor injection: rewrite defs in nested scopes

zerbina committed:
implement code generation for module destructors

The module de-initialization procedures (also refered to as module
destructors) are a bit tricky to handle. We only know their contents
*after* we know all alive globals, but thanks to pure globals, we only
know all alive globals after all alive procedures were discovered and
processed.

Mirroring the original logic, we finish the module destructors right
after wrapping up the main part of code generation for the main module.
Same as with the original logic, this means that new globals discovered
past that point will not be considered for destruction (or
initialization).

zerbina committed:
collectors: re-parent the symbols

Turning the top-level code into a procedure requires changing the owner
of all the definitions; later passes (such as `lambdalifting`) that
inspect the owner of symbols would get confused otherwise.

zerbina committed:
tests: enable `tlobaldestructors.nim` for the VM

zerbina committed:
remember the module for global destructor calls

The information is going to be required for emitting the call into
the correct module destructor.

zerbina committed:
collectors: don't re-parent iterators

This is, unfortunately, also a required fix at the moment, at least
until the C code generator treats the procedure like any other.

zerbina committed:
collectors: cache the positions of important modules

zerbina committed:
backends: restore the previous module initialization order

Make the initialization order the same as the one previously used for
the C target. Whether that's a good idea is questionable, but changing
the order later on is easy, now that it's decided in a single place.

zerbina committed:
cgen: restore setup of emulated TLS

The `initThreadVarsEmulation` call is now emitted at the end of
`systems`'s data-init procedure. This is effectively the same as
calling it in-between `system`'s data-init and module-init procedure,
but has the benefit of not requiring orchestrator support for now.

zerbina committed:
IC: make the IC backend work again

For simplicity, the IC backend now acts as a front-end for the C code-
generation orchestrator, being responsible for turning the rodfile
contents into a `ModuleList`.

While this is going to make further changes to `cgen` much easier, as
the IC backend no longer has to be special cased as much, it has the
significant downside of always having to run code generation again,
even for modules that did not change.

zerbina committed:
vmbackend: remove dead code

zerbina committed:
cgen: reword a comment

zerbina committed:
cgen: remove the unused `pending` parameter

The associated comment is also removed.

zerbina committed:
collectors: small grammar fix

zerbina committed:
Merge branch 'devel' into unify-backend-top-level-statement-processing

zerbina commented on compiler/backend/ccgstmts.nim:
```diff
@@ -156,40 +156,31 @@ proc genGotoVar(p: BProc; value: PNode) =
-  let valueAsRope = potentialValueInit(p, v, value)
   if sfGlobal in v.flags:
+    #assert {sfPure, sfThread} * v.flags != {}, "normal globals don't reach here"
+    # XXX: thanks to ``.compileTime`` globals, the assertion cannot be used
+    #      yet
```
I'm going to try disallowing their usage outside of compile-time-only context once this PR is merged.

The naive approach of detecting illegal usages during the first semantic pass didn't work out -- the detection probably needs to happen during the second pass (`sempass2`).

zerbina commented on compiler/backend/cgen.nim:
```diff
@@ -1663,24 +1603,16 @@ proc shouldRecompile(m: BModule; code: Rope, cfile: Cfile): bool =
 
-proc writeModule(m: BModule, pending: bool) =
+proc finalizeMainModule*(m: BModule) =
+  generateThreadVarsSize(m) # TODO: not the job of the code generator
```
It's still done in the code generator for now, but will eventually become the responsibility of the orchestrator.

zerbina committed:
Merge branch 'devel' into unify-backend-top-level-statement-processing

zerbina commented on compiler/backend/cgen.nim:
```diff
@@ -1706,11 +1638,9 @@ proc finalCodegenActions*(graph: ModuleGraph; m: BModule; n: PNode) =
-    if {optGenStaticLib, optGenDynLib, optNoMain} * m.config.globalOptions == {}:
-      for i in countdown(high(graph.globalDestructors), 0):
-        n.add graph.globalDestructors[i]
+    if emulatedThreadVars(m.config) and m.config.target.targetOS != osStandalone:
+      # raise the dependency on behalf of ``genDatInitCode``:
+      discard cgsym(m, "initThreadVarsEmulation")
```
This kind of dependency raising is all going to be phased out.

Ideally, the bits around generating the thread-local storage setup/teardown logic are all implemented in a backend-agnostic way, so that they can be re-used by the VM and JS backends (once they support threading, that is).

zerbina commented on compiler/ic/cbackend.nim:
```diff
@@ -183,22 +163,51 @@ proc generateCode*(g: ModuleGraph) =
+      let n = unpackTree(g, pos, m.fromDisk.topLevel, p)
+      discard pass.process(c, n)
+
+    # XXX: the order in which the modules are closed is incorrect
```
A pre-existing issue. The IC mechanism doesn't record the module closed order, so the order in which modules are registered with the file manager is used.

zerbina commented on compiler/mir/mirbridge.nim:
```diff
@@ -80,6 +82,89 @@ proc echoOutput(config: ConfigRef, owner: PSym, body: PNode) =
+        if hasInput(body, Operation i):
+          # the definition has a starting value
+          # XXX: consider disallowing inputs to 'def's, they complicate things
+          #      quite a bit (as made evident here)
```
I'm still a bit torn on whether or not to disallow this. While `def`s supporting input makes rewriting them harder (which is the reason for `mirgen` emitting a `def` + `init` for locations that have lifetime hooks), it, by construction, prevents the definition from being separated from its initial assignment.

saem replied:
I would leave them together for now. There might be some value in having more specialized node kinds, such as with initialization, no-initialization, and default initiatlization (something like that). Encoding the decision like that upfront might take out some of the querying/conditionals mixed into the transformation logic.

zerbina commented on compiler/mir/mirbridge.nim:
```diff
@@ -80,6 +82,89 @@ proc echoOutput(config: ConfigRef, owner: PSym, body: PNode) =
+        elif sym.typ.kind in {tyVar, tyLent}:
+          # XXX: this works around an issue in ``transf``, where initialization
+          #      of the view is disjoint from its definition. Fix the bug and
+          #      then remove this special case
```
The issue is fixed by #744.

zerbina commented on compiler/sem/collectors.nim:
```diff
@@ -1,17 +1,31 @@
+## instead of reading the modules' content from the rodfiles, it's collected
+## via the pass interface.
+
+# TODO: the module needs a better name
```
Suggestions for better names are welcome -- I haven't been able to come up with one yet.

saem replied:
- `modulelowering`
- `moduleloweringpass`

zerbina commented on compiler/sem/collectors.nim:
```diff
@@ -46,18 +99,310 @@ type
+    for it in n.items:
+      changeOwner(it, newOwner)
+
+proc setupModule*(graph: ModuleGraph, idgen: IdGenerator, m: PSym,
```
The procedure might need a more descriptive name. While it does set up a `Module` instance, it also does a lot more.

zerbina committed:
collectors: fix destruction order of globals

They were destroyed in the order they were defined, while it needs to
happen in the reverse.

zerbina committed:
collectors: work around semantic analysis issue

Semantic analysis sometimes marks temporaries as globals, which is
incorrect -- temporaries are just local variables inserted by the
compiler; they must never outlive their defining scope.

`mirgen` partially sanitizes this already (temporaries coming out of
`astgen` never have the `sfGlobal` flag set), but they still must not
be registered with the module structs.

zerbina committed:
cbackend: transform module-init procedures as procedures

Their bodies were previously still treated as top-level statements, but
that's just a bad idea, even if only meant as an interim solution.

zerbina committed:
cgen: fix `.global`s not being initialized

A bug local to this PR. If the initializer is sufficiently simple, the
logic in `genSingleVar` still expected `assignGlobalVar` to make sure
that the global was initialized. This isn't the case anymore, so the
globals were left uninitialized.

In addition to removing `potentialValueInit`, `assignGlobalVar` is also
removed, as it does nothing else besides dispatching to
`definedGlobalVar`.

zerbina committed:
mirbridge: fix initialization of globals

Just removing the `def` in case it doesn't have an initializer is wrong.
Globals defined in loops and globals for which the zero-representation
is not valid (e.g. those that have type fields) need to be explicitly
set to their default value.

In order to also support modules being set-up and destroyed multiple
times during the lifetime of a program (not possible right now, but
might be in the future), assigning the default value is performed for
all globals.

There were also two other bugs with the rewriting logic:
- `mnkDefUnpack` and `mnkDefCursor` weren't considered
- the `mnkInit` source argument use the `arg` mode instead of the
  `consume` mode

zerbina committed:
jsgen: prevent duplicate 'init' procedure definitions

zerbina committed:
properly check for globals defined in procedures

Only testing for the `sfPure` flags misses threadvars, and is now
replaced with testing wether the symbol's owner is not a module.

zerbina committed:
jsgen: fix incorrect stack-traces (part 2)

Remove the insertion of stack-trace entries for top-level code. The only
top-level code that reaches the JavaScript code generator are routine
definitions and the 'main' procedure contents, both which don't need
any stack-trace handling inserted (this also reduced the amount of
JavaScript code that is generated).

In addition, the temporary `PProc`s created for top-level code no longer
need any dedicated option handling, so `initProcOptions` is obsolete.

zerbina committed:
jsbackend: don't emit empty module-init procedures

zerbina committed:
jsgen: fix incorrect stack-traces (part 1)

All routines defined in the outermost scope have the `sfGlobal` flag
set, so the flag cannot be used to mark a procedure as the module-init
procedure. Instead, the `sfMainModule` flags is now reused to signal
this.

zerbina committed:
std/sysrand: fix `unrandom` when using NodeJS

Emit statements at the module level can no longer be used to introduce
globals.

zerbina committed:
tests: turn a test failure into a knownIssue

Variables (read: globals) defined in statement lists that are later
lifted into lambda expressions, stopped working at all and now always
either crash the code generator or lead to illegal code being
generated.

zerbina committed:
tests: fix `js/tclosures.nim`

The test relied on emitting top-level JavaScript functions via the `asm`
statement, something that is no longer possible.

It now uses a mixture of `exportc` and `importc` in order to simulate
an imported procedure.

zerbina committed:
tests: enable `tglobals.nim` for the VM target

While not enforced through a shared code path, the VM backend now emits
a call to the pre-initialization procedure in the correct place, making
the test succeed.

zerbina committed:
cgen: restore `.compileTime` globals

With the changes to `genSingleVar`, each usage of a `.compileTime` led
to it being initialized again, which, for obvious reasons, is incorrect.

zerbina committed:
transf: fix incorrect re-parenting of globals

The copied globals still need to have the same owner as the source
symbol. This important for module-level globals, because if they're
re-parented to the module init procedure, `extractGlobals` would pick
them up as globals defined inside routines.

zerbina committed:
mirbridge: work around `transf` issue

zerbina committed:
Merge branch 'devel' into unify-backend-top-level-statement-processing

zerbina committed:
mirbridge: work around yet another issue with globals

It is legal to use the `.global` pragma for module-level definitions;
`once` does so, for example. We keep this working for now by explicitly
skipping them during global def rewriting, so that the JS code generator
can "properly" support them.

zerbina committed:
std/sysrand: use `importc` instead of `importjs`

zerbina committed:
mirbridge: remove the view-related workaround

The bug is gone now that the `devel` branch was merged.

zerbina committed:
tests: fix header file issues

The C code generator previously generated no code at all (not even the
includes) for globals that are using the `header` pragma and don't have
an initializer expression. This is incorrect, and is now fixed,
requiring one to really only import the modules available for ones
targeted platform.

zerbina committed:
test: make the `js/tclosures.js` test work

zerbina committed:
jsgen: don't default initialize pure globals

Default initializing them breaks the `global === undefined` mechanism.
Yet another workaround for `jsgen` not extracting pure globals...

zerbina committed:
(refactor): clean up the comments

* remove stale comments
* fix typos and grammar issues
* clarify some

zerbina committed:
threads: fix emulated TLS

The `globalSlot` was reset when executing the module initialization
procedure for the `system` module, which caused undefined behaviour.
It's now marked as `.noInit`, and `mirbridge` is adjusted to respect the
flag.

zerbina committed:
tests: fix the `tstrict_funcs_imports` test

Only include the `registry` module when building for Windows.

saem commented on compiler/ast/ast_query.nim:
```diff
@@ -103,6 +103,9 @@ const
   routineDefs* = declarativeDefs + {nkMacroDef, nkTemplateDef}
   procDefs* = nkLambdaKinds + declarativeDefs
   callableDefs* = nkLambdaKinds + routineDefs
+  entityDefs* = callableDefs + {nkIdentDefs, nkVarTuple, nkConstDef, nkForStmt}
```
I like it.

saem commented on tests/stylecheck/tstyle_imports.nim:
note to self: this changed because the `.headers` pragma now works correctly.

saem commented on tests/misc/tlifted_lambda_issue.nim:
```diff
@@ -0,0 +1,28 @@
+proc sp(cb: proc())=
+  cb()
+
+sp:
```
Ah, definite bug in sem, I think it might be easier to fix now that I dropped `nfBlockArgs`. Although, thinking about it a bunch of small fixes might be required before the behaviour is quite right. I'll see if I can poke at it after I've wrapped up some other sem changes.

zerbina replied:
The elif branch in `sigmatch.paramTypesMatchAux` that uses `matchesVoidProc` is where lifting a statement list into a lambda expression happens.

saem replied:
Crap, I lost my reply. :( Shortened version, of what I recall.

I think the issue is `prepareOperand` needs to store more than one analysis per argument, because semanticizing for a formal parameter that's lifted to a procedural isn't the same as say `typed`. So we should do the conversion earlier and store the conversion for the arg in its list of analysed possibilities. Then depending upon which one we choose to dispatch, we pick that analysis. Something like `seq[array[AnalysisKind, PNode]]`, where `seq.len == args.len` and `AnalysisKind` covers, concrete, iterator, meta, typed, procedural, untyped, etc.

The thing is shadow scopes juggling becomes bonkers, we might need to create them as nested based on unique sequence of analyses. So we can clearly discard things.

saem commented on lib/std/sysrand.nim:
```diff
@@ -83,7 +83,9 @@ when defined(js):
-    {.emit: "const _nim_nodejs_crypto = require('crypto');".}
+    proc require(p: cstring): JsRoot {.importc.}
+
+    var crypto_module {.exportc: "_nim_nodejs_crypto".} = require("crypto")
```
```suggestion
    let crypto_module {.exportc: "_nim_nodejs_crypto".} = require("crypto")
```

Not that it should make too much of a difference, but it's the right idea as it _should_ be an immutable ref.

saem commented on compiler/sem/collectors.nim:
```diff
@@ -46,18 +99,306 @@ type
+proc registerGlobals(stmts: seq[PNode], structs: var ModuleStructs) =
+  ## Create an entry in `structs` for each global or threadvar defined at the
+  ## the module level. `stmts` represents the imperative part of a module's
+  ## body.
```
```suggestion
  ## module level (within the module imperative body `stmts`).
```

saem commented on compiler/sem/collectors.nim:
```diff
@@ -46,18 +99,306 @@ type
   result = move ModuleListBackend(graph.backend).modules
   graph.backend = nil
 
+proc distribute(n: PNode, decl, imperative: var seq[PNode]) =
```
```suggestion
proc classify(n: PNode, decl, imperative: var seq[PNode]) =
```

An alternative name, might be better. Either that or `bucket` or `group`; `distribute` doesn't indicate the aggregation or collecting aspect.

zerbina replied:
Yeah, `distribute` I was very unsure about -- I think I like `group` most.

saem commented on compiler/sem/collectors.nim:
```diff
@@ -20,23 +34,62 @@ import
+    globals*: seq[PSym]
+      ## all top-level globals part of the module
+
+    globals2*: seq[PSym]
```
```suggestion
    nestedGlobals*: seq[PSym]
```

Soft suggestion, might not be worth it consider the field might go away eventually.

saem commented on compiler/sem/collectors.nim:
```diff
@@ -20,23 +34,62 @@ import
+    structs*: ModuleStructs
+      ## the contents of the module's structs
+
+    # the module-bound procedures. Conceptually, these operate on the module
```
```suggestion
    # the module-bound procedures. Conceptually, these operate on the module's
```

saem commented on compiler/mir/mirbridge.nim:
```diff
@@ -80,6 +82,87 @@ proc echoOutput(config: ConfigRef, owner: PSym, body: PNode) =
+            buf.add MirNode(kind: mnkInit)
+        elif {sfImportc, sfNoInit} * sym.flags == {} and
+             {lfDynamicLib, lfNoDecl} * sym.loc.flags == {}:
+          # XXX: ^^ re-think this condition from first principles. Right now,
+          #      it's just meant to make some tests work
```
These seem like "uncontrolled" locations, having said that, I suspect sem isn't quite doing enough here.

saem reviewed (approved):
This is awesome! All my comments are really minor, but there is lots of great consolation and things are quite a bit simpler.

We're on the cusp of a module type, too. Lots of exciting stuff. 🤩

zerbina commented:
Thank you for the review, @saem!

zerbina commented:
I just noticed that I had missed removing the TODO regarding the module renaming -- that's fixed now.

zerbina committed:
address review comments

- rename the `collectors` module to `modulelowering`
- rename `globals2` to `nestedGlobals`
- rename `distribute` to `group`
- fix typos and grammar mistakes
- make `sysrand.crypto_module` a `let` global

Thank you, @saem.

saem commented:
/merge

I think that's everything.
